text/0000-wallet-replay-protection.md

```diff
@@ -0,0 +1,298 @@
+
+- Title: `wallet-(re)play-protection`
+- Authors: Phyro, John Tromp, Yeastplume
+- Start date: Aug 9, 2020
+- RFC PR: Edit if merged: [mimblewimble/grin-rfcs#0000](https://github.com/mimblewimble/grin-rfcs/pull/0000) 
+- Tracking issue: [Edit if merged with link to tracking github issue]
+
+---
+
+# Summary
+[summary]: #summary
+
+A few months ago, a new class of attacks was found on Grin that we call (re)play attacks. A Mimblewimble transaction that already happened can be replayed if the exact conditions are recreated. In order to be able to replay a transaction the following conditions must be true:
+1. The inputs must be in the utxo set
+2. The outputs of the transaction must not be in the utxo set (Grin does not allow duplicate outputs in the utxo set)
+
+This means that if Alice sent some coins to Bob and the outputs have been spent, anyone that saw their original transaction could replay the transaction if the same inputs existed in the utxo set. But why would the same inputs exist on the chain in the first place? While this seems harmless at first, it can with some creativity and careful coordination be used take someone else's coins without their permission. The attack is not easy to pull off, but is doable in some scenarios. Similarly, a play attack comes from the same reasoning, but with the difference that a transaction never made it to the chain for some reason. One such reason could be that an input was spent before it was broadcasted which would make the current transaction invalid. In this document, we propose new wallet behaviours that make it robust in the face of (re)play attacks so the end users can't be victims of these attacks.
+
+# Motivation
+[motivation]: #motivation
+
+The goal of this RFC is to propose new wallet rules that, when strictly followed, protect the user from all known malicious (re)play attacks. This is done by changing the transaction building process but at the same time keeping it configurable enough to still allow the user to take control of their privacy. The solution mostly supports all the transaction building flows that were used before while encouraging default use of PayJoin transaction with some probability. Contributing inputs on to a 'receive' transaction can be turned off in the wallet configuration settings while still keeping the protection against replay attacks. The wallet also allows choosing different wallet configurations for different use cases e.g. a withdrawal from an exchange never does a PayJoin transaction. 
+
+# Community-level explanation

```

Comment:
Try to write the community level explanation even more high level, and without any diagrams, or naming conventions (i.e. Transaction `T` etc as this is technical). Instead try to explain what's being implemented on the wallet with words, from an end user's perspective.

For example: 
* We can stop the attacks by making impossible to recreate the transactions that created the outputs.
* We do so by creating an output that is never being spent. 
* Outputs that are created at the same time as this one, cannot be replayed.

Reply:
Thanks for the input! Do you think it would be a good idea to move the diagrams to the beginning of the `Reference-level explanation` or just throw them out? I include them because I find it easier to see what's going on, but it could be that I'm just used to them because I've done quite a few of these. Are they any help to helping you guys understand the idea or not really?

Reply:
Diagrams in general are helpful, I think! Could make sense to move them to reference level 👍